libsplinter/src/node_registry/yaml/remote.rs

```diff
@@ -0,0 +1,337 @@
+// limitations under the License.
+
+//! A remote, read-only node registry.
+//!
+//! This module contains the [`RemoteYamlNodeRegistry`], which provides an implementation of the
+//! [`NodeRegistryReader`] trait.
+//!
+//! [`RemoteYamlNodeRegistry`]: struct.RemoteYamlNodeRegistry.html
+//! [`NodeRegistryReader`]: ../../trait.NodeRegistryReader.html
+
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::{Duration, Instant};
+
+use openssl::hash::{hash, MessageDigest};
+
+use crate::hex::to_hex;
+use crate::node_registry::{MetadataPredicate, Node, NodeRegistryError, NodeRegistryReader};
+
+use super::{validate_nodes, LocalYamlNodeRegistry};
+
+/// A remote, read-only node registry.
+///
+/// The `RemoteYamlNodeRegistry` provides access to a remote node registry YAML file over HTTP(S).
+/// The remote registry file must be a YAML sequence of nodes, where each node is valid (see
+/// [`Node`] for validity criteria). Read operations are provided by the [`NodeRegistryReader`]
+/// implementation.
+///
+/// The remote YAML file is cached locally by saving it to the filesystem. This ensures that the
+/// registry will remain available even if the remote file becomes unreachable. The on-disk
+/// location of the local cache is determined by the `cache_dir` argument of the registry's
+/// [`constructor`].
+///
+/// On initialization, the `RemoteYamlNodeRegistry` will attempt to immediately fetch and cache the
+/// remote file. If this fails, the registry will log an error message and attempt to fetch/cache
+/// the remote file every time a read query is made on the registry (through one of the
+/// [`NodeRegistryReader`] methods) until the file is successfully cached. Until the registry has a
+/// local cache, it will behave as an empty registry. Once the remote file has been successfully
+/// cached for the first time, the registry will always provide data from the cache.
+///
+/// If an `automatic_refresh_period` is provided to the registry's [`constructor`], the registry
+/// will attempt to automatically refresh the cache in the background after the given time since
+/// the last automatic refresh attempt has elapsed.
+///
+/// If a `forced_refresh_period` is provided to the registry's [`constructor`], the registry will
+/// attempt to cache the remote file on each read after the given time since the last successful
+/// cache attempt has elapsed.
+///
+/// If a forced or automatic cache refresh fails for any reason, an error message will be logged
+/// and the previously cached registry values will continue to be used. The next time the registry
+/// is read, it will try again to refresh the cache.
+///
+/// [`Node`]: struct.Node.html
+/// [`NodeRegistryReader`]: trait.NodeRegistryReader.html
+/// [`constructor`]: struct.RemoteYamlNodeRegistry.html#method.new
+pub struct RemoteYamlNodeRegistry {
+    internal: Arc<Mutex<Internal>>,
+    _automatic_refresh_thread_handle: Option<thread::JoinHandle<()>>,
```

Comment:
This will need to be used/dealt with in order to stabilize

Reply:
Removed in favor of a shutdown handle